src/environment.cpp

```diff
@@ -681,20 +658,22 @@ typedef struct {
 				continue;
 			v3s16 p = p0 + block->getPosRelative();
 
-			if (!m_aabms[c])
+			if (!m_aabms[c]) {
+				if (block->content_only)
+					return;
```

Comment:
if no abm for 'c'  and block contain only 'c' - no need to scan whole block.
good for air (~50% of world at ~0 height).